[WIP, RFC] drivers/eeprom: Common API to access EEPROM memory

maribu opened the pull request:
### Contribution description

This PR sketches how a generic EEPROM API could look like and be implemented. The user-facing API is mostly identical to the API in `periph/eeprom.h` in current master.

### Testing procedure

In this early stage testing is not yet possible. Once the API has been agreed on, EEPROM drivers could be adapted to this API and then be tested.

### Issues/PRs references

In PR https://github.com/RIOT-OS/RIOT/pull/11929 a driver for external I2C EEPROM is suggested. The need for a common EEPROM API instead of a duct-tape wrapper over the existing EEPROM API became apparent to allow using external and internal EEPROMs with a consistent API

benpicco commented on drivers/eeprom/include/eeprom_params.h:
```diff
@@ -0,0 +1,74 @@
+        .handle = NULL,
+    },
+#endif
+#ifdef MODULE_PERIPH_EEPROM
```
`MODULE_AT24CXXX`

benpicco commented on drivers/include/eeprom.h:
```diff
@@ -0,0 +1,183 @@
+ * @return  The size of the EEPROM device in bytes
+ * @retval  -1              No EEPROM device for parameter @p dev
+ */
+ssize_t eeprom_size(eeprom_t dev);
```
On 16-bit platforms, the EEPROM size might actually be bigger that `ssize_t`

maribu replied:
I changed it to use `size_t`, so that up to 64 KiB can be used on 8-bit and 16-bit platforms. I'm not sure if it makes more sense to use `uint32_t` as the current EEPROM-API does, or the additional overhead on 8-bit and 16-bit platforms is not worth it.

Maybe introducing `eeprom_off_t` for size and positions is a solution, such as:

```C
#ifndef HAVE_EEPROM_OFF_T
typedef size_t eeprom_off_t;
#endif
```

Offloading decisions to the user is always a good idea :-D

benpicco replied:
Well I have here a 1MBit SPI EEPROM, so `size_t` is still too small. In fact I only noticed this because CI would fail on those platforms (I used `size_t` too at first.)

I don't know how big the overhead of just always using `uint32_t` is though.

In my driver I have a configurable [`AT25XXX_PARAM_ADDR_LEN`](https://github.com/RIOT-OS/RIOT/pull/11945/files#diff-0db8ad96013abefa0eb5a2660166ec68R53), we could make it `EEPROM_ADDR_LEN` and set the type of `eeprom_off_t` based on that.
It will make the code a bit more difficult though.

maribu replied:
> I don't know how big the overhead of just always using `uint32_t` is though.

Likely less than the overhead of the indirect function call via the `eeprom_driver_t` (but I honestly don't know and would have to disassemble the code to see). I added the `eeprom_off_t` and it defaults to `uin32_t`. Less maintenance headache by using the `uint32_t` is likely worth the overhead. And having a `typedef` in place, this could still be optimized later on without changing the API.

aabadie commented on drivers/eeprom/include/eeprom_driver.h:
```diff
@@ -0,0 +1,150 @@
+     * pointer is `NULL`, a fallback implementation that calls
+     * @ref eeprom_driver_t::write is used instead.
+     */
+    int (*set)(void *handle, size_t pos, uint8_t val, size_t len);
```
set and erase function are not strictly needed since they just require the write method. Maybe we could simply the driver struct with just init/read/write functions ?

benpicco replied:
Some EEPROMs do offer opcodes to do page-wise erase, might me faster than manually overwriting all the bytes.

maribu replied:
That's why the `eeprom_driver_t` has an `erase()` function pointer. Only if that is `NULL`, the fallback implementation of using `set()` is used.

(The global `erase()` function is intended to be used by the user to erase memory. Maybe extending the doc to point to that would be a good idea.)

aabadie replied:
ok for erase but this is not the case for set since it can be implemented using the write function (like clear).

maribu replied:
done

maribu commented:
>     * Now one has to explicitly pass the device used if there are multiple EEPROM available. This changes quite a bit the existing API and will require user code to be updated.

Like for every other RIOT-API. Citing the [Guide for writing a device driver in RIOT](https://github.com/RIOT-OS/RIOT/wiki/Guide%3A-Writing-a-device-driver-in-RIOT):

> Additionally towards ease of use, all device drivers in RIOT should provide a similar 'look and feel'.

> * To remove the need of the extra `eeprom_t`, how about only use one EEPROM at a time ? An external one being used instead of the CPU one if present ?

Citing again from the RIOT driver guide:

> Third, it should always be possible, to handle more than a single device of one kind. Drivers and their interfaces are thus designed to keep their state information in a parameterized location instead of driver defined global variables.

So this is against RIOT's policy.

>     * Another solution could be to "merge" all available EEPROM into a single global memory ?

This is super complex. Image one has three EEPROM devices. Now the order of the initialization is changed --> the memory layout is messed up. Or one of the three fails to initialize. What will the memory layout look like? There more I think about it the more pitfall and corner cases pop up.

aabadie commented on drivers/eeprom/include/eeprom_params.h:
```diff
@@ -0,0 +1,74 @@
+    },
+#endif
+#ifdef MODULE_AT24CXXX
+    {
```
How about this (list of) struct is taken from a macro defined in AT24CXXX params header ? This would allow to configure and use multiple external EEPROM of the same kind. The default driver params could also be overloaded the usual way.

maribu replied:
I like the idea. I implemented it a bit different to allow only changing only the part of the array related to a single driver. E.g. if I had 2 AT25*, one AT24C* and one internal EEPROM device, this would allow to overwrite the AT25* related part of the array.

aabadie commented:
> Citing the Guide for writing a device driver in RIOT

Make sense. I just wanted to discuss possible alternatives.
The current design makes it little more complex for the calling code which has to deal with available EEPROM source.

benpicco commented on drivers/include/eeprom.h:
```diff
@@ -0,0 +1,189 @@
+ */
+static inline int eeprom_clear(eeprom_t dev, eeprom_off_t pos, size_t len)
+{
+    return eeprom_set(dev, pos, 0, len);
```
Some EEPROMs do provide pagewise erase.

benpicco commented on drivers/eeprom/include/eeprom_driver.h:
```diff
@@ -0,0 +1,128 @@
+ * @brief   Driver interface for EEPROM devices
+ *
+ * This interface must be implemented by every EEPROM device driver. The
+ * function pointers in @ref eeprom_driver_t::set and
```
> ```warning: unable to resolve reference to `eeprom_driver_t::set' for \ref command```

maribu replied:
Oh yes, I removed that function. Thanks for noticing!

Fixed :-)

benpicco commented on drivers/eeprom/include/eeprom_driver.h:
```diff
@@ -0,0 +1,128 @@
+     *
+     * This function is optional to implement by the driver. If the function
+     * pointer is `NULL`, a fallback implementation that calls
+     * @ref eeprom_driver_t::set is used instead.
```
> ```warning: unable to resolve reference to `eeprom_driver_t::set' for \ref command```

maribu replied:
Fixed :-)

maribu commented on drivers/eeprom/include/eeprom_driver.h:
```diff
@@ -0,0 +1,128 @@
+     * @retval  0       Failed to initialize the device
+     * @return          The size of the device in bytes
+     */
+    eeprom_off_t (*init)(void *handle, const void *params);
```
@aabadie: You suggested to drop the `params` here, right?

maribu commented on drivers/eeprom/include/eeprom_driver.h:
```diff
@@ -0,0 +1,128 @@
+    const eeprom_driver_t *driver;  /**< The driver of the EEPROM device */
+    const void *params;             /**< The initialization parameters */
+    void *handle;                   /**< The handle to pass to the driver */
+} eeprom_dev_t;
```
@aabadie: I think if we drop the `params` in the init callback, it will no longer be possible to add multiple external EEPROM devices that are using the same driver, without some dark magic. Assume the init function of an external driver is only called with the device handle to initialize. With what parameters should it initialize that device handle?

If there is only one device of that kind, it is clear. But RIOT policy is to allow multiple devices of a kind. I don't see how this could work.

benpicco commented:
What's the verdict on this one?

maribu commented:
Some other PRs are waiting for a common EEPROM API. I think that the basic API provided here - basically the `periph_eeprom` API extended to have a device descriptor - should do fine.

I'm willing to provide PRs to adapt any existing code in RIOT master to the new API. I also provide a PR as fall back wrapper for the existing API, that simple uses the first EEPROM device available - that way external applications could work without being updated.

But that all should be done once a common API is agreed upon.

benpicco commented:
I'm fine with this API.
If the concern is that this adds too much overhead/unneeded indirection because nobody will ever use more than one EEPROM at a time, I can also draft a header-only api (with some common support code) like what we have for the watchdog.

benpicco commented:
> I also provide a PR as fall back wrapper for the existing API, that simple uses the first EEPROM device available - that way external applications could work without being updated.

This could also include a fast path to directly (or with a thin wrapper) call the `periph_eeprom` functions if no external EEPROM modules are used.

gschorcht commented on drivers/eeprom/eeprom.c:
```diff
@@ -0,0 +1,170 @@
+
+void eeprom_init(void)
+{
+    for (unsigned i = 0; i < EEPROM_NUMOF; i++) {
```
As long as you have the dummy device, this would lead to crashes.
```suggestion
    for (unsigned i = 0; i < EEPROM_NUMOF - 1; i++) {
```

maribu replied:
The dummy is not intended to stay. The problem is, that this API will not compile if the list of backends is empty. This is actually very nice to have, as this is preferable over an error at runtime. However, I wanted that reviewers can compile this during review.

gschorcht replied:
That means, the `eeprom` module will not be compilable without having any real device? It is ok if there is no test case where this could happen.

maribu replied:
It is basically only a wrapper allowing to access multiple EEPROM devices with the same API. Without EEPROM drivers present, there is no point in having this API. So I'd say it should not compile without any backend, as this will warn users early on that a driver is missing in `USEMODULE`.

gschorcht replied:
ok

gschorcht commented on drivers/eeprom/eeprom.c:
```diff
@@ -0,0 +1,170 @@
+void eeprom_init(void)
+{
+    for (unsigned i = 0; i < EEPROM_NUMOF; i++) {
+        _sizes[i] = eeprom_devs[i].driver->init(eeprom_devs[i].handle,
```
We should use
```
        assert(eeprom_devs[i].driver && eeprom_devs[i].handle)
```
here and in all other functions that use these pointers to avoid silent crashes.

maribu replied:
Hmmm. I just thought that some drivers might not want an handle. E.g. for a periph device, the number of device is fixed. Maybe that driver prefers to have the `handle` as `NULL` and handle the device state locally, as it is safe to statically allocate a fixed amount of memory for that specific case.

But the `assert()` for the driver is a very good idea. Thanks!

gschorcht replied:
> Hmmm. I just thought that some drivers might not want an handle.

OK, that's a good argument. If the driver insists on a valid handle, the driver has to use `assert` for the handle.

aabadie replied:
I think we should decouple underlying device params from the common eeprom init function. See how this is done with netdev.
I would also use an `eeprom_t dev` parameter for the init function.

maribu replied:
So providing a `setup()` taking the parameters that and an `init()` function? I never fully understood why that was done, as it increase ROM size (both by having two functions to call, and by having two times function boilerplate instructions needed to comply with the caller convention), increases the complexity (now two functions have to be called instead of one), increases the sources for errors (the order of the function calls matters), and has no apparent benefit.

aabadie replied:
This way you get rid of all macros from `eeprom_params.h` and let the user deal with driver params setup.
This also has the benefit of not having to modify the generic API (and params) when a new external EEPROM driver is added to RIOT.

maribu replied:
> This way you get rid of all macros from `eeprom_params.h`

I wouldn't call it getting rid of, as the macros would only be moved to different headers.

> This also has the benefit of not having to modify the eeprom_dev source code (and params) when a new external EEPROM driver is added to RIOT.

How about not having to add them in `eeprom_params.h`, and let `eeprom_params.h` just `#include` those. That would be pretty much the same result.

Another advantage of the current approach is that `EEPROM_NUMOF` is evaluated by the compiler. With the netdevs, the user has to count the number of network devices himself. This is definitely a task better left to computers than to humans.

gschorcht replied:
I agree with @aabadie. The EEPROM API should be free from device specific stuff. That would also imply that the API only defines the type `eeprom_dev_t` for the device list `eeprom_devs`. The device list `eeprom_devs` would be defined by the board definition.

maribu replied:
What about users that use external EEPROM, that is not part of the board?

maribu replied:
How about moving the initialization out of this API and providing a function to register an eeprom device, similar to the SAUL registry. The initialization could than be performed in auto init with a single init function for each device. It would still not overcome the issue that users have to count the number of EEPROM devices, but it would provide the strict separation of configuration parameters and common API without the overhead of two init functions per device.

gschorcht replied:
They could define such a configuration file in $(APPDIR).  According the the example in PR #12333 there would be a file `eeprom_dev_conf.h` which is included by the `eeprom` module. This file is usualy in board definition but could also be in `$APPDIR` provided that that `$(APPDIR)` the first directory in the `INCLUDES` path.

`EEPROM_NUMOF` would also evaluated during compile time.

gschorcht replied:
> provided that that `$(APPDIR)` the first directory in the `INCLUDES` path.

BTW, at some point I suggested adding `$(APPDIR)` as the first entry to be able to override the default board definition by the application. But that didn't find consensus.

gschorcht commented on drivers/eeprom/include/eeprom_params.h:
```diff
@@ -0,0 +1,92 @@
+#endif
+    /* dummy to get this compiled for now. TODO: Remove this */
+    {
+        .driver = NULL,
```
A possibility to avoid the dummy device being NULL and causing crashes is to introduce a standard "unsupported device", as we do in other drivers, which does nothing else than print a message, that the function is not supported by the device.

maribu replied:
See explanation above.

maribu commented:
OK. I split out initialization of this PR entirely and added a function `void eeprom_register(const eeprom_driver_t *driver, void *handle, eeprom_off_t size);` to registers devices to the common API instead. I think this could be a good compromise.

aabadie commented on drivers/eeprom/include/eeprom_types.h:
```diff
@@ -0,0 +1,139 @@
+/**
+ * @brief   Offset / size type in EEPROM devices
+ */
+typedef uint32_t eeprom_off_t;
```
What about calling the type `eeprom_offset_t` and be explicit ? I find `eeprom_off_t` a bit misleading.

aabadie replied:
Same for `HAVE_EEPROM_OFFSET_T` macro.

maribu replied:
Done

aabadie commented on drivers/include/eeprom.h:
```diff
@@ -0,0 +1,194 @@
+#endif
+
+/* Include implementation of the static inline functions */
+#include "eeprom_implementation.h"
```
Is there a real benefit with the static inline functions ? Is not I would just implement declare them here and implement them in `eeprom.c`. Or maybe keep them static inline but put them in `eeprom.h`.

It's just that I don't like having to include a header at the end of another one so it's better if we could avoid this.

maribu replied:
> Is there a real benefit with the static inline functions?

It depends. In any case, it will be faster. In some cases, the compiler will be able to optimize out (some of) the range checks done there.

Regarding ROM size, this could be both a minor improvement or a minor regression. It will depend on the number of places where the common API is called and on how much of the code can be optimized out. In general, I expect small applications to be slightly small with this, and other being slightly bigger.

I have no strong feelings about this; so I'd just as happy move it back to `eeprom.c`.

@benpicco suggested to reduce the overhead here, maybe he wants to comment before I move that back?

> It's just that I don't like having to include a header at the end of another one so it's better if we could avoid this.

This could actually also be included at the top of the file, if no compiler flags are set to warn about late declarations.  But I think the C standard is fine with multiple or late declarations (as long as declarations and the implementations match in name and signature). The advantage of still having the declaration here is that the generated documentation will be properly grouped. And the advantage of moving the implementation out is the separation of API and implementation.

This is by the way also not uncommon, e.g. [`xtimer.h`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/xtimer.h) includes the static inline implementations at the end of the header.

aabadie commented on drivers/eeprom/include/eeprom_types.h:
```diff
@@ -0,0 +1,139 @@
+ *
+ * This function is to be called only once and during boot.
+ */
+void eeprom_init(void);
```
Not needed anymore.

maribu replied:
Indeed. And additionally it was in the wrong header.

Thanks for spotting this. I removed it.

aabadie commented on drivers/eeprom/eeprom.c:
```diff
@@ -0,0 +1,72 @@
+#include <errno.h>
+#include <string.h>
+
+#include "assert.h"
```
This is a system include, so should be `#include <assert.h>` and moved with the other system includes above.

gschorcht replied:
I have to contradict that unfortunately. This header is from `core/includes` and not from of standard C libraies. IMHO, it is correct to include it with `"assert.h"`.

maribu replied:
You are both right. On POSIX systems it is a system header. But the standard requests that a failing assert send the signal SIGABORT (not sure about the spelling), so the standard version is not available on embedded systems. Luckily RIOT provides a compatible implementation (form the user's perspective; there is no SIGABORT used for obvious reasons, though).

With that in mind, it is not that clear if it should be treated as system lib or not. I've seen both versions in RIOT. I'm personally fine with either.

maribu replied:
So, if no one speaks up I just keep it as it is.

aabadie commented on drivers/eeprom/eeprom.c:
```diff
@@ -0,0 +1,72 @@
+        return -EFAULT;
+    }
+
+    uint8_t buf[EEPROM_SET_BUF];
```
```suggestion
    uint8_t buf[EEPROM_SET_BUF] = { 0 };
```
?

maribu replied:
That's better! Done

gschorcht commented on drivers/eeprom/include/eeprom_types.h:
```diff
@@ -0,0 +1,139 @@
+     * @retval  0       Failed to initialize the device
+     * @return          The size of the device in bytes
+     */
+    eeprom_off_t (*init)(void *handle, const void *params);
```
Has this still to be part of the driver. If I'm right, a device that is registered has to be initialized by the application or `board_init` function. That would be different, if you would define a static list of devices which are initialized by the `init` function of this API.

maribu replied:
No, it is a leftover I overlooked

maribu replied:
Removed

gschorcht commented:
> I like the current approach.

I still don't really like the approach of registration. Usually in RIOT all hardware components are defined by static initializers. Why not use the same for the configuration of existing EEPROM devices? If `_eeprom_devs[]` would be declared by the board or the application like following
```c
static const eeprom_dev_t _eeprom_devs[] = {
    {
      .driver = abc_eeprom_driver,
      .handle = &abc_eeprom,
      .params = &abc_eeprom_params,
      .size = ABC_EEPROM_SIZE
    },
    {
      .driver = xyz_eeprom_driver,
      .handle = &xyz_eeprom,
      .params = &xyz_eeprom_params,
      .size = XYZ_EEPROM_SIZE
    },
};
```
the `init` function of the API could iterate over all EEPROM devices and could call 
```.c
    _eeprom_driver(dev)->init(_eeprom_handle(dev), _eeprom_params(dev));
```
The init function could be called in `sys/auto_init`.

This approach would have the advantages
- that initialization of all EEPROMs would be done automatically
- the user has not take what functions to call and when
- the `EEPROM_NUM_OF` would be simply `ARRAY_SIZE(_eeprom_devs)

benpicco commented:
FYI: [Some EEPROM](http://ww1.microchip.com/downloads/en/devicedoc/Atmel-8807-SEEPROM-AT24MAC402-602-Datasheet.pdf)s provide an EUI-48 or EUI-64.

maribu commented:
> I still don't really like the approach of registration.

Well, this is basically the trade-off we have to make. Either we pull in implementation details of every implementation to allow for static initialization, or we keep the API completely distinct from those details. Recently, you and @aabadie convinced me that the latter is to be preferred in order to trade initialization overhead for better maintainability and sound architectural design.

> Usually in RIOT all hardware components are defined by static initializers.

E.g. with SAUL and network device drivers we have the registration for the same reason.

> If `_eeprom_devs[]` would be declared by the board or the application

This would have its own disadvantage; namely that board configurations get longer and that end users plugging external stuff (not included in their board) would have more effort to configure.

I have however no strong feeling about this decision. Maybe @aabadie and @benpicco  could give their opinion on whether they prefer registration or `static` initializers? I would than just do as the majority says. (Luckily there are three of you, so there should be no draw :smile:)

benpicco commented:
If the registration approach helps keep the board files cleaner I'm all for it.
And if we manage to resist the temptation to compare `EEPROM_NUMOF` in processor macros we can even populate it automatically:

```C
#define EEPROM_NUMOF (MCU_EEPROM + AT24XXX_NUMOF + AT25XXX_NUMOF …)
```

benpicco commented:
@gschorcht What about EEPROMs that are integrated into the MCU? It would be quite bad if every board would have to declare them.

gschorcht commented:
> @gschorcht What about EEPROMs that are integrated into the MCU? It would be quite bad if every board would have to declare them.

For boards that have same properties, we use `boards/common/*/board_common.h` for the definitions.

But that's not my point. All hardware configurations in RIOT are done as static const definitions during the compile time. There is only one exception, the netif. However, we are still looking for an approach on how to get the static definition of GNRC_NETIF_NUMOF and the dynamic Netif instantiation consistent.

This will be the same here. We have the static definition of `EEPROM_NUMF` and the dynamic registration of the devices that really there. Also your macro hack wouldn't solve this problem since it doesn't cover the case that a board could have multiple eeproms of same type.

The attempt of a dynamic instantiation without a dynamic registry just has its limits :worried:

To be clear, I'm not against the approach. It just wanted to tell my concerns. There are a lot of other use case where such a registration would help, e.g., for GPIO extenders.

benpicco commented:
> For boards that have same properties, we use `boards/common/*/board_common.h` for the definitions.

That doesn't help with your static array example.
Say you have an ATmega board with an additional I2C EEPROM.
You can't have a split array between common code and the board, so either you have to declare both EEPROMs in the board code (meh) or you do something like

```
#define EXTERNAL_EEPROMS            \
    {                               \
      .driver = i2c_eeprom_driver,  \
      .params = &i2c_eeprom_params, \
      .size = I2C_EEPROM_SIZE       \
    },

…

#define INTERNAL_EEPROMS                \
    {                                   \
      .driver = atmega_eeprom_driver,   \
      .params = NULL,                   \
      .size = ATMEGA_EEPROM_SIZE        \
    },

…

static const eeprom_dev_t _eeprom_devs[] = {
    INTERNAL_EEPROMS 
    EXTERNAL_EEPROMS 
};
```

Or just have a separate struct for external EEPROMs - that should work too.

> Also your macro hack wouldn't solve this problem since it doesn't cover the case that a board could have multiple eeproms of same type.

If you have 
```C
#define AT24XXX_NUMOF ARRAY_SIZE(at25xxx_params)
```
it should work.

benpicco commented:
I'm wondering: Why not just use the existing MTD API for EEPROMs?

maribu commented:
Any idea why MTD exposes the concept of sectors and pages? The API seems not to be using this. IMHO it would be nice to get rid of these details, so that MTD could be applied more broadly. But to me it seems that indeed that API should work just fine.

Some work would be needed to update users like the EEPROM Registry, so that they could be used with external EEPROM. (And any other memory that has an MTD API.)

Btw: Wouldn't storage would be the correct term rather than memory? To me, storage is persistent, while memory is volatile.

maribu commented:
Just using the MTD API is clearly the better approach. No need to reinvent the wheel.
